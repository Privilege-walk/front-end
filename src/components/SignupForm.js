import React, { useState } from 'react';
import { Formik } from 'formik';
import axios from 'axios';
import { Navigate } from "react-router-dom";
import * as yup from 'yup';

import Form from "react-bootstrap/Form";
import {
    Label, 
    Button, 
    FormGroup, 
    FormFeedback,
    Input,
    Row,
    Col
} from 'reactstrap';
import { restClient } from '../api/restInterceptor';




const schema = yup.object().shape({
    username: yup.string().required("Username is required"),
    password: yup.string().required("Password is required"),
    firstName: yup.string().required("First Name is required"),
    lastName: yup.string().required("Last Name is required"),
    email: yup.string().required("Email is required").email("Invalid email address")
});


export default function SignUpForm({onSubmit,setFirstPage}) {
    const [login, setLogin] = useState(false);
    const [errMsg, setErrMsg] = useState("");

    function submitSignupForm(formData){
        const data = {
            "first_name": formData["firstName"],
            "last_name": formData["lastName"],
            "username": formData["username"],
            "password": formData["password"],
            "email": formData["email"]
        };
        restClient.post("/auth/signup/", data).then(async res => {
            if(res.data.created == "success"){
                setLogin(true);
                setFirstPage('login');
            }else{
                setErrMsg(res.data.created);
            }
        })
    }

    function handleSubmit(formData){
        setErrMsg("");
        
        if(onSubmit){
            onSubmit(formData);
        }else{
            submitSignupForm(formData);
        }
        
    }

    function onLogin() {
        setFirstPage('login');
    }

    return (
        <div className="Form">
        <Formik
            validationSchema={schema}
            onSubmit={handleSubmit}
            initialValues={{
                firstName: '',
                lastName: '',
                username: '',
                password: '',
                email: '',
              
              }}
        >
            {({
                handleSubmit,
                handleChange,
                values,
                touched,
                valid,
                errors
            }) => (
                <Form onSubmit={handleSubmit} className="mt-4">
                    
                    <Row className="mb-1">
                        <FormGroup
                        md="6"
                        className="position-relative"
                        >
                            <Label htmlFor="firstName">First Name</Label>
                            <Input
                                type="text"
                                id="firstName"
                                name="firstName"
                                value={values.firstName}
                                onChange={handleChange}
                                valid={touched.firstName && !errors.firstName}
                                invalid={errors.firstName?true:false}
                            />
                            <FormFeedback type="invalid">
                                {errors.firstName}
                            </FormFeedback>
                        </FormGroup>
                    </Row>


                    <Row className="mb-1">
                        <FormGroup
                        md="6"
                        className="position-relative"
                        >
                            <Label htmlFor="lastName">Last Name</Label>
                            <Input
                                type="text"
                                id="lastName"
                                name="lastName"
                                value={values.lastName}
                                onChange={handleChange}
                                valid={touched.lastName && !errors.lastName}
                                invalid={errors.lastName?true:false}
                            />
                            <FormFeedback type="invalid">
                                {errors.lastName}
                            </FormFeedback>
                        </FormGroup>
                    </Row>

                    <Row className="mb-1">
                        <FormGroup
                        md="6"
                        className="position-relative"
                        >
                            <Label htmlFor="username" >Username</Label>
                            <Input
                                type="text"
                                id="username"
                                name="username"
                                value={values.username}
                                onChange={handleChange}
                                valid={touched.username && !errors.username}
                                invalid={errors.username?true:false}
                            />
                            <FormFeedback type="invalid">
                                {errors.username}
                            </FormFeedback>
                        </FormGroup>
                    </Row>

                    <Row className="mb-1">
                        <FormGroup
                        md="6"
                        className="position-relative"
                        >
                            <Label htmlFor="password">Password</Label>
                            <Input
                                id="password"
                                type="password"
                                name="password"
                                value={values.password}
                                onChange={handleChange}
                                valid={touched.password && !errors.password}
                                invalid={errors.password?true:false}
                            />
                            <FormFeedback type="invalid">
                                {errors.password}
                            </FormFeedback>
                        </FormGroup>
                    </Row>

                    <Row className="mb-1">
                        <FormGroup
                        md="6"
                        className="position-relative"
                        >
                            <Label htmlFor="email" >Email</Label>
                            <Input
                                type="text"
                                id="email"
                                name="email"
                                value={values.email}
                                onChange={handleChange}
                                valid={touched.email && !errors.email}
                                invalid={errors.email?true:false}
                            />
                            <FormFeedback type="invalid">
                                {errors.email}
                            </FormFeedback>
                        </FormGroup>
                    </Row>

                    <Row data-testid="signupErrorId" className="mb-3 errMsg justify-content-center">{errMsg}</Row>

                    {/* <Row className="justify-content-center my-2"> */}
                    <Button size="lg" type="submit" 
                    >
                        Sign Up
                    </Button>
                    <Button size="lg"
                        style={{marginLeft: 20}}
                        onClick={onLogin}
                    >
                        Login
                    </Button>
                    {/* </Row> */}
                    
                </Form>
            )}
        </Formik>

        {
            login &&
            <Navigate to="/" />
        }
        </div>
    )
}