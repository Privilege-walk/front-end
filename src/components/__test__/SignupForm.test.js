// Validate input data before sending. if invalid post request should not be sent.
// username: field cannot be empty.
// password: field cannot be empty. has to be more than 8 characters
// first_name: field cannot be empty.
// Last name: field cannot be empty.
// Email: field cannot be empty, has to have @ sign.


// successful creation. should show some output on the frontend.
// username exists
// Email exists.
import SignupForm from '../SignupForm';

const setup = () => {
    let utils;
    act (() => {
        utils = render(<SignupForm />);
    }); 
    let data = {
        "username": "user1",
        "password": "password1",
        "firstName": "Jane",
        "lastName": "Doe",
        "email": "jane.doe@privilegewalk.com"
    }
    let inputs = {};
    for(let key in data.keys()){
        inputs[key] = screen.getByLabelText(key);
    }
    const signupBtn = screen.getByText('Sign Up', {selector: 'button'});
    return {
        data,
        inputs,
        signupBtn,
        ...utils
    }
}

test('renders without crashing',  ()=> {
    const utils = setup(); 
    const { asFragment, inputs } = utils;
    expect(asFragment()).toMatchSnapshot();
    for(let key in inputs.keys()){
        expect(inputs[key].value).toBe('');
    }
});

test('Successful sign up', () => {
    const utils = setup(); 
    const { inputs, data, signupBtn } = utils;
    var mock = new MockAdapter(axios);
    mock.onPost(
        'http://localhost:8000/auth/login/', 
        data
    ).reply(200, {created: "success"});

    for(let key in inputs.keys()){
        fireEvent.change(inputs[key], {target: {value: data[key]}});
    }

    await act(async () => {   
        await fireEvent.click(signupBtn);
    });
})

