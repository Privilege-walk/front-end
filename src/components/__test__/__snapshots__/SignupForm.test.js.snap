// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="Form"
  >
    <form
      class="mt-4"
    >
      <div
        class="mb-1 row"
      >
        <div
          class="position-relative mb-3"
          md="6"
        >
          <label
            class="form-label"
            for="firstName"
          >
            First Name
          </label>
          <input
            aria-invalid="false"
            class="form-control"
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
            type="invalid"
          />
        </div>
      </div>
      <div
        class="mb-1 row"
      >
        <div
          class="position-relative mb-3"
          md="6"
        >
          <label
            class="form-label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            aria-invalid="false"
            class="form-control"
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
            type="invalid"
          />
        </div>
      </div>
      <div
        class="mb-1 row"
      >
        <div
          class="position-relative mb-3"
          md="6"
        >
          <label
            class="form-label"
            for="username"
          >
            Username
          </label>
          <input
            aria-invalid="false"
            class="form-control"
            id="username"
            name="username"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
            type="invalid"
          />
        </div>
      </div>
      <div
        class="mb-1 row"
      >
        <div
          class="position-relative mb-3"
          md="6"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            aria-invalid="false"
            class="form-control"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
            type="invalid"
          />
        </div>
      </div>
      <div
        class="mb-1 row"
      >
        <div
          class="position-relative mb-3"
          md="6"
        >
          <label
            class="form-label"
            for="email"
          >
            Email
          </label>
          <input
            aria-invalid="false"
            class="form-control"
            id="email"
            name="email"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
            type="invalid"
          />
        </div>
      </div>
      <div
        class="mb-3 errMsg justify-content-center row"
        data-testid="signupErrorId"
      />
      <div
        class="justify-content-center my-2 row"
      >
        <button
          class="btn btn-secondary btn-lg"
          type="submit"
        >
          Sign Up
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
